mesh: Various fixes

michal-narajowski opened the pull request:
This PR includes various fixes regarding:
- Provisioning
- Key refresh procedure
- Configuration server
- Transport layer configuration

Commit:
mesh: Fix resending provisioning link ack

The comparison of a current tx link transaction id caused
us to not send ack and in effect fail provisioning.

Signed-off-by: Michał Narajowski <michal.narajowski@codecoup.pl>

Commit:
mesh: Start protocol timer on send message in advertising bearer

Protocol timer was not started when sending a control message
in advertising bearer.

Signed-off-by: Michał Narajowski <michal.narajowski@codecoup.pl>

Commit:
mesh: Add reset link functionality to provisioning bearers

Reset provisioning bearers when bt_mesh_reset() is called. Accept
another provisioning attempt after link close.

Signed-off-by: Michał Narajowski <michal.narajowski@codecoup.pl>

Commit:
mesh: Fix key refresh procedure

Key refresh procedure was ignored on non-primary subnet.

Mesh Profile Specification v1.0.1:
3.10.4 Key Refresh procedure
"This procedure is used when the security of one or more network
keys and/or one or more of the application keys has been compromised
or could be compromised."

"It is possible to update each NetKey independently of all other
NetKeys. A Key Refresh procedure for one NetKey can be in a different
phase to another Key Refresh procedure for other NetKeys."

Signed-off-by: Michał Narajowski <michal.narajowski@codecoup.pl>

Commit:
mesh: Fix Vendor Model Subscription Get procedure

Invalid argument was passed to bt_mesh_model_tree_walk.

Signed-off-by: Michał Narajowski <michal.narajowski@codecoup.pl>

Commit:
mesh: Adjust config values to pass Transport test cases

These values were derived experimentally by running PTS test cases.

Signed-off-by: Michał Narajowski <michal.narajowski@codecoup.pl>

Commit:
mesh: Fix net_buf leak in advertising provisioning bearer

Unsegmented buffer in bearer_ctl_send was not unrefed causing
a memory leak.

Signed-off-by: Michał Narajowski <michal.narajowski@codecoup.pl>

jhedberg commented on subsys/bluetooth/mesh/pb_adv.c:
```diff
@@ -188,11 +188,20 @@ static void reset_adv_link(void)
+		/* Accept another provisioning attempt */
+		link.id = 0;
+		atomic_clear(link.flags);
+		link.rx.id = 0x7F;
+		link.tx.id = 0xFF;
```
I'm starting to think we might want a define for this magic `0x7f` since it seems quite prevalent in the code.

michal-narajowski replied:
Agreed. I added a new commit with the define.

jhedberg commented on subsys/bluetooth/mesh/beacon.c:
```diff
@@ -334,6 +334,13 @@ static void secure_beacon_recv(struct net_buf_simple *buf)
 
 	cache_add(data, sub);
 
+	kr_change = bt_mesh_kr_update(sub, BT_MESH_KEY_REFRESH(flags), new_key);
+	if (kr_change) {
```
I wonder if I've then read the spec wrong (or perhaps the spec was changed?) about reacting to non-primary subnets. It might make sense to include a spec section reference in the commit message, and perhaps also a quote from the text there.

michal-narajowski replied:
I added a quote from the spec in the commit message.

jhedberg commented on subsys/bluetooth/mesh/pb_adv.c:
```diff
@@ -538,6 +538,8 @@ static void send_reliable(void)
 			bt_mesh_adv_send(buf, &buf_sent_cb, NULL);
 		}
+
+		net_buf_unref(buf);
```
Shouldn't `link.tx.buf[i]` be set to NULL as well, since it seems that's the reference you're unrefing here?

trond-snekvik replied:
Agreed, I think this change is wrong, the buffers should be freed once the transaction has been acked in gen_prov_ack and link_ack, freeing them here removes the ability to retransmit on a nack.

michal-narajowski replied:
bt_mesh_adv_send() will call net_buf_ref every time a buffer is sent so if we retransmit and only unref (and then set it to NULL) in free_segments(), we are leaking them.

trond-snekvik replied:
But the ref in bt_mesh_adv_send will be unref'd in adv.c:adv_send, won't it? And the new unref on line 631 is to cancel out the adv_buf_create above it, or am I missing something here

michal-narajowski replied:
Yes, you are right. I overdid it. :smile: So only one unref was missing.

jhedberg commented on subsys/bluetooth/mesh/pb_adv.c:
```diff
@@ -593,6 +595,8 @@ static void prov_retransmit(struct k_work *work)
 			bt_mesh_adv_send(buf, &buf_sent_cb, NULL);
 		}
+
+		net_buf_unref(buf);
```
Same thing here with `link.tx.buf[i]`

Commit:
mesh: Do not ignore message with DST as RFU group address

According to Mesh Profile Test Specification we should still
relay messages with DST address that is RFU.

Fixes MESH/NODE/RLY/BV-01-C.

Signed-off-by: Michał Narajowski <michal.narajowski@codecoup.pl>

Commit:
mesh: Replace hardcoded duplicate cache size

The hardcoded value was arbitrary and interfered with message cache
mechanism causing invalid behavior and PTS test to fail.

Fixes MESH/NODE/RLY/BV-02-C.

Signed-off-by: Michał Narajowski <michal.narajowski@codecoup.pl>

Commit:
mesh: Add a define representing max transaction id

This value was used quite often in the file so adding a
descriptive name should improve readability.

Signed-off-by: Michał Narajowski <michal.narajowski@codecoup.pl>

trond-snekvik commented on subsys/bluetooth/mesh/net.c:
```diff
@@ -88,7 +88,7 @@ struct bt_mesh_net bt_mesh = {
 };
 
-static uint32_t dup_cache[4];
+static uint32_t dup_cache[CONFIG_BT_MESH_MSG_CACHE_SIZE];
```
The dup cache is not the same as the message cache, but it's an interesting change. 

In theory, we can replace the entire message cache with this, and save a lot of deobfuscation, although it'll emit false positives for one out of every `4.29 billion / cache_size` packets. I believe this was the original motivation behind having a "proper" message cache, but it seems like an acceptable downside to me. I can make a separate PR after this one if @jhedberg agrees

michal-narajowski replied:
Yes, I understand that it's not the same. I was hoping for a discussion about this change. Initially I though this was left out by accident and I wanted to remove it since we have proper message cache. Then I ran the test with this small change and it fixed the test so I figured we will talk about it in the comments.

trond-snekvik replied:
Interesting. The intent is that this would filter out most duplicate messages, so we don't have to run deobfuscation over and over again, only to discover that the message is a duplicate. Sounds like there's a hole in the message cache implementation then, and this change hides it. The way I see it, we should either replace the message cache completely or fix it. I'm leaning towards the former 😄

jhedberg replied:
It seems like this discussion wasn't yet resolved? What's the value for `CONFIG_BT_MESH_MSG_CACHE_SIZE` that you're using? I'd like to understand better why any of this is causing problems, since the purpose of `dup_cache` is just to eliminate unnecessary processing, i.e. the stack should work whether it exists or not, and regardless of what size it is.

jhedberg replied:
That said, obviously there is a risk for collisions since we're hashing a message into just 32 bits, however I'd like to understand better why the test case was failing.

michal-narajowski replied:
I looked at it once more and I think I messed up a bit. At first I was testing with CONFIG_BT_MESH_MSG_CACHE_SIZE=2 because that was the default value in PTS IXIT. It failed (more messages were ignored than PTS expected) so I made this change and it worked. Then I tested with size=10 and it worked again. Now I tried with size=10 but without the fix and it works. So it works when size >= dup_cache size which is understandable. Still I think that we should have a define or a config value for dup_cache size as this magic value (4) can be confusing.

trond-snekvik replied:
I see. I think the collision risk with the 32bit hash is acceptable, given that we can completely avoid doing deobfuscation on duplicate packets. Even at an rx rate of 100 new packets per second and a cache size of 10, we'll be dropping a valid packet only once every 50 days. The average packet drop rate from network interference is orders of magnitude higher, so it wouldn't be noticeable in the application layer. 

I think the current change is the right one, but as it makes the message cache redundant, we should additionally remove that completely.

trond-snekvik replied:
Discussed with @jhedberg offline:
Okay to remove the existing message cache functionality, and rely entirely on the dup_cache. 
@michal-narajowski I think this should be done as part of this PR, as it already makes the existing message cache functionality redundant.

Commit:
mesh: Ignore Friend Request from local interface

Friend Request is also transmitted over local queue and
if both Friend and LPN features are enabled then we would
try to establish friendship with ourselves.

Fixes MESH/NODE/FRND/LPN testcases.

Signed-off-by: Michał Narajowski <michal.narajowski@codecoup.pl>
